meta-mbl-distro/scripts/lib/mbl/update/payloadbuilder.py

```diff
@@ -0,0 +1,173 @@
+# Copyright (c) 2019 Arm Limited and Contributors. All rights reserved.
+#
+# SPDX-License-Identifier: BSD-3-Clause
+
+"""Provides an abstract factory for creating objects for building payloads."""
+
+import abc
+import pathlib
+
+import mbl.update.appsimage as appsimage
+import mbl.update.bootimage as bootimage
+import mbl.update.payloadarchiver as payloadarchiver
+import mbl.update.rootfsimage as rootfsimage
+import mbl.update.wksbootloaderslotimage as wksbootloaderslotimage
+import mbl.util.tinfoilutil as tutil
+
+
+class PayloadBuilder(abc.ABC):
+    """Abstract factory for creating objects used when buliding payloads."""
+
+    def __init__(self, tinfoil):
+        """
+        Construct a PayloadBuilder object.
+
+        This should only be called by subclasses and is a utility for saving
+        data required when instantiating the objects that the factory provides.
+
+        Args:
+        * tinfoil Tinfoil: BitBake Tinfoil object.
+
+        """
+        self._tinfoil = tinfoil
+        self._deploy_dir = pathlib.Path(
+            tutil.get_bitbake_conf_var("DEPLOY_DIR_IMAGE", tinfoil)
+        )
+
+    @staticmethod
+    def for_format_version(payload_format_version, tinfoil):
+        """
+        Create a PayloadBuilder object.
+
+        The returned object is a factory that provides objects to create
+        payloads of the given payload format version.
+
+        Args:
+        * payload_format_version int: payload format version for which the
+          objects created by the factory should build.
+        * tinfoil Tinfoil: BitBake Tinfoil object.
+
+        """
+        if payload_format_version == 1:
+            return PayloadBuilderV1(tinfoil)
+        elif payload_format_version == 3:
+            return PayloadBuilderV3(tinfoil)
+        else:
+            bb.fatal(
+                'Unsupported payload format version "{}"'.format(
+                    payload_format_version
+                )
+            )
+
+    @abc.abstractmethod
+    def create_archiver(self):
+        """Create a PayloadArchiver."""
+        ...
```

Comment:
We shouldn't need these three dots. Is this to appease one of the insanity checkers?

Reply:
We need _something_ there to satisfy Python itself, either `1`, `pass`, `...` or whatever. Otherwise Python 3.6 says something like: "IndentationError: expected an indented block ".

The [abc module docs](https://docs.python.org/3/library/abc.html) all use `...`, so I did too.

Reply:
Oh, does the docstring count as the "_something_" I need there?

Reply:
Yes, it does, cool, I'll remove the `...`s.

Reply:
https://github.com/ARMmbed/meta-mbl/pull/781/commits/3c4543350777adf76dabf3bbe77995ec95162a18

Reply:
yeah the docstring counts as something. I think in python 2.7 you had to have `pass` or something else also, but that changed for python3 IIRC.